### ISSUE TITLE
Concurrent calls to addReadStream results in empty files

### ISSUE THREAD

**Issue by tepez** [thejoshwolfe/yazl#8]
This standalone code demonstrates the issue.
`generateFile` returns a readable stream to a file. We use it to create files concurrently (`concurrency` determines how many in parallel).

If `concurrency` is 1 - all files are written, but if we use 2 or higher some file will be empty:

```js
var Yazl = require('yazl'), Bluebird = require('bluebird'),
  Stream = require('stream'), Fs = require('fs');

var zipFile = new Yazl.ZipFile(),
  zipStream = zipFile.outputStream;

var concurrency = 5;

console.log('concurrency = ' + concurrency);

// return a readable stream to 1000 bytes file
var generateFile = function(idx) {
  var stream = new Stream.Readable();
  for (var i = 0; i < 1000; i++) {
    stream.push('' + idx);
  }
  stream.push(null);
  return stream;
};

Bluebird.map([ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 ], function(idx) {

  var fileStream = generateFile(idx);

  zipFile.addReadStream(fileStream, idx + ".txt", { compress: false });

  var totalWritten = 0;
  fileStream.on('data', function(chunk) { totalWritten += chunk.length; });

  var deferred = Bluebird.defer();

  fileStream.on('error', function(err) {
    console.log('fileStream error');
    deferred.reject(err);
  });
  fileStream.on('end', function() {
    console.log("fileStream end: " + totalWritten + " written");
    deferred.resolve();
  });

  return deferred.promise.delay(0);

}, { concurrency: concurrency }).then(function() {
  zipFile.end(function(total) { console.log("zipFile end cb: " + total + " written"); });
}).catch(function(err) {
  zipStream.emit('error', err);
});

var total = 0;

zipStream.on('data', function(chunk) { total += chunk.length });
zipStream.on('end', function() { console.log("zipStream end: " + total + " written"); });

zipStream.pipe(Fs.createWriteStream('tmp.zip'));
```

```
>node tmp.js
concurrency = 1
fileStream end: 1000 written 
fileStream end: 1000 written 
fileStream end: 1000 written 
fileStream end: 1000 written 
fileStream end: 1000 written 
fileStream end: 1000 writtez
fileStream end: 1000 written
fileStream end: 1000 written
fileStream end: 1000 written
fileStream end: 1000 written
zipFile end cb: 11042 written
zipStream end: 11042 written

>node tmp.js
concurrency = 2
fileStream end: 1000 written 
fileStream end: 1000 written 
fileStream end: 1000 written 
fileStream end: 1000 written 
fileStream end: 1000 written 
fileStream end: 1000 written 
fileStream end: 1000 written 
fileStream end: 1000 written 
fileStream end: 1000 written 
fileStream end: 1000 written 
zipFile end cb: 6042 written 
zipStream end: 6042 written
 
>node tmp.js
concurrency = 5
fileStream end: 1000 written 
fileStream end: 1000 written 
fileStream end: 1000 written 
fileStream end: 1000 written 
fileStream end: 1000 written 
fileStream end: 1000 written 
fileStream end: 1000 written 
fileStream end: 1000 written 
fileStream end: 1000 written 
fileStream end: 1000 written 
zipFile end cb: 3042 written 
zipStream end: 3042 written                
```

![image](https://cloud.githubusercontent.com/assets/3265649/6266500/22f3f904-b843-11e4-91bf-a2485619c4e7.png)

**Comment by thejoshwolfe** [comment]
I've reproduced your issue (sorta). The `concurrency > 1` cases for me look more like this:

```
$ node issue8.js 
concurrency = 2
fileStream end: 1000 written
fileStream end: 1000 written
fileStream end: 1000 written
fileStream end: 1000 written
fileStream end: 1000 written
fileStream end: 1000 written
fileStream end: 1000 written
fileStream end: 1000 written
fileStream end: 1000 written
fileStream end: 1000 written
zipFile end cb: -1 written
```

but the `concurrency === 1` case looks the same as yours. There's clearly something going wrong here.

**Comment by thejoshwolfe** [comment]
Here's a clue: if you comment out both calls to `.on('data', ...)`, then the resulting zip file is correct for all concurrency levels (but you lose your byte count logging, of course).

I think this has something to do with "flowing mode" vs "non-flowing mode" streams in node 0.10, and possibly also related to "old style" streams. Are you using node 0.10? Supposedly node 0.12 did something that may have made this not a problem. I'm not really sure how it all works, but from what I know about node 0.10's stream api "on('data') is bad, and pipe() is good". That's how yazl was designed, and apparently the two do not get along. I'll add a note in the readme to this effect.

If you're interested in counting the bytes that flow through a stream, yazl has a ByteCounter class in index.js that does exactly that by using pipe() instead of on('data'). You can copy it to your code for equivalent functionality. (it's only 9 lines, and likely the simplest possible non-trivial example of extending the `Transform` class.)

**Comment by thejoshwolfe** [comment]
see cc7d74cb51f192ed68bb4e815143de0a25340769

**Comment by tepez** [comment]
@thejoshwolfe I confirm that the test case now works for me as well (running node 11.13)
I'm actually not interested in the bytes count. I added the `on('data')` calls just for debugging.

Unfortunately for me removing the `on('data')` calls does not solve the issue in my real app, where I still have a situation where creating a zip file only work when concurrency=1.
I'll dig in further and reopen this issue if needed.

Thanks for quick response!

**Comment by thejoshwolfe** [comment]
Another data point: [groovebasin](https://github.com/andrewrk/groovebasin) uses `addFile` concurrently with no limit to download multiple files as a zip, and the code has been working for several months of regular use with no observed problems.

https://github.com/andrewrk/groovebasin/blob/dfd0fd8b9b6af5f6b857bb26128bd3c043487e09/lib/server.js#L299
